opm/simulators/wells/WellConvergence.cpp

```diff
@@ -0,0 +1,155 @@
+/*
+  Copyright 2017 SINTEF Digital, Mathematics and Cybernetics.
+  Copyright 2017 Statoil ASA.
+  Copyright 2016 - 2017 IRIS AS.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+#include <opm/simulators/wells/WellConvergence.hpp>
+
+#include <opm/simulators/timestepping/ConvergenceReport.hpp>
+#include <opm/simulators/utils/DeferredLoggingErrorHelpers.hpp>
+#include <opm/simulators/wells/WellInterfaceGeneric.hpp>
+#include <opm/simulators/wells/WellState.hpp>
+
+namespace Opm
+{
+
+void WellConvergence::
+checkConvergenceControlEq(const WellState& well_state,
+                          const Tolerances& tolerances,
+                          const double well_control_residual,
+                          ConvergenceReport& report,
+                          DeferredLogger& deferred_logger) const
+{
+    double control_tolerance = 0.;
+    using CR = ConvergenceReport;
+    CR::WellFailure::Type ctrltype = CR::WellFailure::Type::Invalid;
+
+    const int well_index = well_.indexOfWell();
+    const auto& ws = well_state.well(well_index);
+    if (well_.wellIsStopped()) {
+        ctrltype = CR::WellFailure::Type::ControlRate;
+        control_tolerance = tolerances.rates; // use smaller tolerance for zero control?
+    }
+    else if (well_.isInjector() )
+    {
+        auto current = ws.injection_cmode;
+        switch(current) {
+        case Well::InjectorCMode::THP:
+            ctrltype = CR::WellFailure::Type::ControlTHP;
+            control_tolerance = tolerances.thp;
+            break;
+        case Well::InjectorCMode::BHP:
+            ctrltype = CR::WellFailure::Type::ControlBHP;
+            control_tolerance = tolerances.bhp;
+            break;
+        case Well::InjectorCMode::RATE:
+        case Well::InjectorCMode::RESV:
+            ctrltype = CR::WellFailure::Type::ControlRate;
+            control_tolerance = tolerances.rates;
+            break;
+        case Well::InjectorCMode::GRUP:
+            ctrltype = CR::WellFailure::Type::ControlRate;
+            control_tolerance = tolerances.grup;
+            break;
+        default:
+            OPM_DEFLOG_THROW(std::runtime_error, "Unknown well control control types for well " << well_.name(), deferred_logger);
```

Comment:
We should include `<stdexcept>` to bring a declaration of `std::runtime_error` into scope.

Reply:
done.